packages/webrtc-core/src/media/track.ts

```diff
@@ -0,0 +1,198 @@
+/* eslint-disable no-underscore-dangle */
+import { EventMap } from 'typed-emitter';
+import { EventEmitter } from '../event-emitter';
+import { logger } from '../util/logger';
+import { MediaStreamTrackKind } from '../peer-connection';
+
+export enum Events {
+  Ended = 'ended',
+  Muted = 'muted',
+  /**
+   * Fires when the published state of a LocalTrack changes.  A value of 'true' indicates the
+   * track has been published, and 'false' indicates that it is no longer published.  Tracks are
+   * unpublished by default, and must be published explicitly (no event will be fired to indicate
+   * the initial state of 'unpublished').
+   */
+  PublishedStateUpdate = 'published-state-update',
+  /**
+   * Fires when there has been a change in the underlying track.
+   */
+  UnderlyingTrackChange = 'underlying-track-change',
+}
+
+export interface TrackState {
+  id: string;
+  label: string;
+  muted: boolean;
+}
+
+export type TrackEndEvent = {
+  trackState: TrackState;
+};
+
+export type TrackMuteEvent = {
+  trackState: TrackState;
+};
+// eslint-disable-next-line no-shadow
+export enum TrackStatus {
+  ENDED = 'ended',
+  LIVE = 'live',
+}
+
+// eslint-disable-next-line no-shadow
+export enum TrackKind {
+  AUDIO = 'audio',
+  VIDEO = 'video',
+}
+
+export type TrackPublishEvent = {
+  isPublished: boolean;
+  trackState: TrackState;
+};
+
+export interface TrackEvents extends EventMap {
+  [Events.Ended]: (event: TrackEndEvent) => void;
+  [Events.Muted]: (event: TrackMuteEvent) => void;
+  [Events.UnderlyingTrackChange]: () => void;
+}
+
+export type TrackEffect = unknown;
+// TBD: Fix this once types are published separately
+// export type TrackEffect = BaseMicrophoneEffect | BaseCameraEffect;
+
+/**
+ * Basic Track class.
+ */
+export abstract class Track extends EventEmitter<TrackEvents> {
+  static Events = Events;
+
+  ID: string;
+
+  kind: MediaStreamTrackKind;
+
+  status: TrackStatus;
+
+  label: string;
+
+  #mediaStreamTrack: MediaStreamTrack;
+
+  #mediaStreamTrackWithEffect?: MediaStreamTrack;
+
+  /**
+   * Constructor for the Track class. Creates an empty CoreLocalTrack or uses an existing one.
+   *
+   * @param track - Media stream track.
+   */
+  constructor(track: MediaStreamTrack) {
+    super();
+    this.ID = track.id;
```

Comment:
do we need this if we already have a getter for id that returns the underlying track's id?
and kind of a similar question for label and kind - maybe we should just have getters for them? is there some benefit of having the copies?

Reply:
good question .

we can put all 3 as getter or keep the ID as uuid because it could change on device change or adding effects !?

any thoughts @bbaldino

Reply:
yeah i don't think we need extra state if we already have the track and can retrieve the id from there

Reply:
plus the getter below doesn't even seem to use this field?

Reply:
changed it to UUID